presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitSource.java

```diff
@@ -221,38 +227,105 @@ public static HiveSplitSource bucketed(
+                    // TODO: Rewrite split source for bucketed mode so we don't have two layers of wrappers. (PerBucket -> BucketedSplits -> AsyncQueue)
+                    class BucketedSplits
+                    {
+                        private final int estimatedQueueSize;
+                        private final Executor executor;
+                        private final List<InternalHiveSplit> discoveredHiveSplits = new ArrayList<>();
+
+                        private AsyncQueue<InternalHiveSplit> queue;
+                        private ListenableFuture<List<ConnectorSplit>> lastBorrowBatchAsyncFuture;
+
+                        public BucketedSplits(int estimatedQueueSize, Executor executor)
+                        {
+                            this.estimatedQueueSize = estimatedQueueSize;
+                            this.executor = requireNonNull(executor, "executor is null");
+
+                            this.queue = new AsyncQueue<>(estimatedQueueSize, executor);
+                        }
+
+                        public synchronized ListenableFuture<?> offer(InternalHiveSplit internalHiveSplit)
+                        {
+                            discoveredHiveSplits.add(internalHiveSplit);
+                            return queue.offer(internalHiveSplit);
+                        }
+
+                        public synchronized ListenableFuture<List<ConnectorSplit>> borrowBatchAsync(int maxSize, Function<List<InternalHiveSplit>, BorrowResult<InternalHiveSplit, List<ConnectorSplit>>> function)
+                        {
+                            checkState(lastBorrowBatchAsyncFuture == null || lastBorrowBatchAsyncFuture.isDone(), "last borrowBatchAsync must finish if exists");
+                            lastBorrowBatchAsyncFuture = queue.borrowBatchAsync(maxSize, function);
+                            return lastBorrowBatchAsyncFuture;
+                        }
+
+                        public synchronized void finish()
+                        {
                             queue.finish();
                         }
-                        return queue;
+
+                        public synchronized boolean isFinished()
+                        {
+                            return queue.isFinished();
+                        }
+
+                        public synchronized int rewind()
+                        {
+                            if (lastBorrowBatchAsyncFuture != null) {
+                                // Caller should no longer need to check the future returned by borrowBatchAsync after calling rewind.
+                                // Explicitly cancel the future as a sanity check.
+                                lastBorrowBatchAsyncFuture.cancel(true);
+                                lastBorrowBatchAsyncFuture = null;
+                            }
+                            int revivedSplitCount = discoveredHiveSplits.size() - queue.size();
```

Comment:
It feels like this number must be computed as the number of finished internal splits at this point.